File: packs/data/aws-dynamodb/scaffold/wiring/python/dynamo_note_repository.py
import os
from datetime import datetime
from typing import Any

import aioboto3

from app.domain.entities.note import Note
from app.domain.interfaces.note_repository import NoteRepository


class DynamoNoteRepository(NoteRepository):
    """DynamoDB implementation of the note repository.

    This file drops into the python-fastapi backend's app/infrastructure/repositories/ unchanged:
    it speaks the same Note entity and satisfies the same NoteRepository interface as the in memory
    reference, so replacing MemoryNoteRepository with it in app/api/deps.py is the whole
    substitution. The table name is read from the environment by default and the endpoint is
    overridable, so one adapter serves both a real regional table and a DynamoDB Local container.
    Every item read back is validated into a Note, so a malformed row fails loudly at this boundary
    rather than silently inside a workflow.
    """

    def __init__(
        self,
        table_name: str | None = None,
        endpoint_url: str | None = None,
        region_name: str | None = None,
    ) -> None:
        resolved = table_name or os.environ.get("NOTES_TABLE_NAME")
        if not resolved:
            raise RuntimeError("NOTES_TABLE_NAME is not set and no table_name was provided")
        self._table_name = resolved
        self._endpoint_url = endpoint_url or os.environ.get("DYNAMODB_ENDPOINT_URL")
        self._region_name = region_name or os.environ.get("AWS_DEFAULT_REGION")
        self._session = aioboto3.Session()

    def _resource(self) -> Any:
        return self._session.resource(
            "dynamodb",
            endpoint_url=self._endpoint_url,
            region_name=self._region_name,
        )

    async def add(self, note: Note) -> Note:
        async with self._resource() as dynamo:
            table = await dynamo.Table(self._table_name)
            await table.put_item(
                Item={
                    "id": note.id,
                    "title": note.title,
                    "body": note.body,
                    "created_at": note.created_at.isoformat(),
                }
            )
        return note

    async def get(self, note_id: str) -> Note | None:
        async with self._resource() as dynamo:
            table = await dynamo.Table(self._table_name)
            response = await table.get_item(Key={"id": note_id})
        item = response.get("Item")
        if item is None:
            return None
        return self._to_note(item)

    async def list_all(self) -> list[Note]:
        async with self._resource() as dynamo:
            table = await dynamo.Table(self._table_name)
            response = await table.scan()
        return [self._to_note(item) for item in response.get("Items", [])]

    @staticmethod
    def _to_note(item: dict[str, Any]) -> Note:
        return Note(
            id=str(item["id"]),
            title=str(item["title"]),
            body=str(item.get("body", "")),
            created_at=datetime.fromisoformat(str(item["created_at"])),
        )

File: packs/data/aws-dynamodb/scaffold/wiring/typescript/src/infrastructure/dynamo-note-repository.ts
import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
import { DynamoDBDocumentClient, GetCommand, PutCommand, ScanCommand } from "@aws-sdk/lib-dynamodb";
import { type Note, noteSchema } from "../domain/note.js";
import type { NoteRepository } from "../domain/note-repository.js";

/**
 * The DynamoDB adapter for the note repository interface. This file drops into the typescript-node
 * backend's src/infrastructure/ unchanged: it imports the same entity and interface from ../domain,
 * so replacing MemoryNoteRepository with it in the composition root is the whole substitution.
 *
 * The table name is read from the environment by default, and the endpoint is overridable, so the
 * one adapter serves both a real regional table and a DynamoDB Local container with no code change.
 * Every item read back is validated into an entity through noteSchema, so a malformed row fails
 * loudly at this boundary rather than silently inside a workflow.
 */
export interface DynamoNoteRepositoryConfig {
  readonly tableName?: string;
  readonly endpoint?: string;
  readonly region?: string;
}

export class DynamoNoteRepository implements NoteRepository {
  readonly #doc: DynamoDBDocumentClient;
  readonly #tableName: string;

  constructor(config: DynamoNoteRepositoryConfig = {}) {
    const tableName = config.tableName ?? process.env.NOTES_TABLE_NAME;
    if (tableName === undefined || tableName === "") {
      throw new Error("NOTES_TABLE_NAME is not set and no tableName was provided");
    }
    this.#tableName = tableName;

    const endpoint = config.endpoint ?? process.env.DYNAMODB_ENDPOINT_URL;
    const region = config.region ?? process.env.AWS_DEFAULT_REGION;
    const client = new DynamoDBClient({
      ...(endpoint !== undefined && endpoint !== "" ? { endpoint } : {}),
      ...(region !== undefined && region !== "" ? { region } : {}),
    });
    this.#doc = DynamoDBDocumentClient.from(client);
  }

  async add(note: Note): Promise<Note> {
    const stored = noteSchema.parse(note);
    await this.#doc.send(
      new PutCommand({
        TableName: this.#tableName,
        Item: {
          id: stored.id,
          title: stored.title,
          body: stored.body,
          createdAt: stored.createdAt.toISOString(),
        },
      }),
    );
    return stored;
  }

  async get(id: string): Promise<Note | null> {
    const result = await this.#doc.send(
      new GetCommand({ TableName: this.#tableName, Key: { id } }),
    );
    return result.Item === undefined ? null : this.#toNote(result.Item);
  }

  async listAll(): Promise<Note[]> {
    const result = await this.#doc.send(new ScanCommand({ TableName: this.#tableName }));
    return (result.Items ?? []).map((item) => this.#toNote(item));
  }

  #toNote(item: Record<string, unknown>): Note {
    return noteSchema.parse({
      id: item.id,
      title: item.title,
      body: item.body,
      createdAt: new Date(String(item.createdAt)),
    });
  }
}
